src/interop/js-interop/js-types.md

```diff
@@ -0,0 +1,242 @@
+---
+title: JS types
+description: Usage information about the core types in JS interop.
+---
+
+Dart values and JS values belong to separate language domains. When compiling to
+Wasm, they execute in separate *runtimes* as well. As such, you should treat JS
+values as foreign types. To provide Dart types for JS values,
+[`dart:js_interop`] exposes a set of [extension types] prefixed with `JS` called
+"JS types".
+
+Importantly, these types have compiler-specific [representation types]. This
+means that their runtime type will differ based on whether code is compiled to
+Wasm or JS. In order to interact with and examine these JS values, you should
+use [`external`] interop members.
+
+## Type hierarchy
+
+JS types form a natural type hierarchy:
+
+- Top type: `JSAny`, which is any non-nullish JS value
+  - Primitives: `JSNumber`, `JSBoolean`, `JSString`
+  - `JSSymbol`
+  - `JSBigInt`
+  - `JSObject`, which is any JS object
+    - `JSFunction`
+      - `JSExportedDartFunction`, which represents a Dart callback that was
+      converted to a JS function
+    - `JSArray`
+    - `JSPromise`
+    - `JSDataView`
+    - `JSTypedArray`
+      - JS typed arrays like `JSUint8Array`
+    - `JSBoxedDartObject`, which allows users to box and pass Dart values
+      opaquely within the same Dart runtime
+
+You can find the definition of each type in the [`dart:js_interop` API docs].
+
+{% comment %}
+TODO (srujzs): Should we add a tree diagram instead for JS types?
+{% endcomment %}
+
+## Conversions
+
+To use a value from one domain to another, you will likely want to *convert* the
+value to the corresponding type of the other domain. For example, you might want
+to convert a Dart `List<JSString>` into a JS array of strings, which is
+represented by the JS type `JSArray<JSString>`, so that you can pass the array
+to a JS interop API.
+
+Dart supplies a number of `extension` members on various Dart types and JS types
+to convert the values between the domains for you.
```

Comment:
Small consideration: maybe we can talk about the fact that they are extension members later. 

For example:
```suggestion
Dart provides a wide variety of members to help you convert values between the two domains.
```

Then when describing the `.toJS` name we can also share that they are added on as extension members. I debate whether we should also say it for `.toDart` (since the member and the type are exposed together, and I could see us one day moving the `.toDart` into the extension type).

For example:
> Helper members to convert from Dart to JS are exposed as extension members on Dart types with names starting with `.toJS`: ...sample-code...
> Conversely, to convert from JS to Dart, you can use members with names starting with `.toDart` on JS types.

Reply:
I think it's a good idea to omit the fact that they are extension members altogether. Mentioning it for the `toJS` case but not the `toDart` case makes it a little confusing as today they're extension members as well.

(I can totally see the motivation for mentioning it though. Trying to use conversion functions with just an import to `package:web` resulting in an error may seem initially confusing.)